connectors/sources/network_drive.py

```diff
@@ -300,6 +411,112 @@ async def get_content(self, file, timestamp=None, doit=None):
+                file_descriptor=file.fd, info=SECURITY_INFO_DACL
+            )
+            return descriptor.get_dacl()["aces"]
+
+    def _dls_enabled(self):
+        if (
+            self._features is None
+            or not self._features.document_level_security_enabled()
+        ):
+            return False
+
+        return self.configuration["use_document_level_security"]
+
+    async def _decorate_with_access_control(self, document, file_path, file_type):
+        if self._dls_enabled():
+            entity_permissions = await self.get_entity_permission(
+                file_path=file_path, file_type=file_type
+            )
+            document[ACCESS_CONTROL] = list(
+                set(document.get(ACCESS_CONTROL, []) + entity_permissions)
+            )
+        return document
+
+    async def _user_access_control_doc(self, user, sid, groups_info, groups_members):
+        prefixed_username = _prefix_user(user)
+        sid_users = _prefix_sid(sid)
+        sid_groups = []
+
+        for group_name, group_sid in groups_info.items():
+            members = groups_members.get(group_name)
+
+            if sid in members.values():
+                sid_groups.append(_prefix_sid(group_sid))
+
+        access_control = [sid_users, prefixed_username, *sid_groups]
+
+        return {
+            "_id": sid,
+            "identity": {
+                "username": prefixed_username,
+                "user_id": sid_users,
+            },
+            "created_at": iso_utc(),
+        } | es_access_control_query(access_control)
+
+    async def get_access_control(self):
+        if not self._dls_enabled():
+            self._logger.warning("DLS is not enabled. Skipping")
+            return
+
+        self._logger.info("Fetching all groups and members")
+        groups_info = await asyncio.to_thread(self.security_info.fetch_groups)
+
+        groups_members = {}
+        for group_name, _ in groups_info.items():
+            groups_members[group_name] = await asyncio.to_thread(
+                self.security_info.fetch_members, group_name
+            )
+
+        self._logger.info("Fetching all users")
+        users_info = await asyncio.to_thread(self.security_info.fetch_users)
+
+        for user, sid in users_info.items():
+            yield await self._user_access_control_doc(
+                user, sid, groups_info, groups_members
+            )
+
+    async def get_entity_permission(self, file_path, file_type):
+        if not self._dls_enabled():
+            return []
+
+        permissions = []
+        if file_type == "file":
+            list_permissions = await asyncio.to_thread(
+                self.list_file_permission,
+                file_path=file_path,
+                file_type="file",
+                mode="rb",
+                access=FilePipePrinterAccessMask.READ_CONTROL,
+            )
+        else:
+            list_permissions = await asyncio.to_thread(
```

Comment:
Did you already look into https://github.com/skelsec/aiosmb ? Seems like this is an async wrapper around smb.

Reply:
Yes, we tried to use the aiosmb module. However it failed to work with MacOS based network drives , so we had to use `smbprotocol`